src/internal/minikubetestenv/deploy.go

```diff
@@ -223,9 +249,13 @@ func pachClient(t testing.TB, pachAddress *grpcutil.PachdAddress, authUser, name
 		if err != nil {
 			return errors.Wrapf(err, "failed to connect to pachd on port %v", pachAddress.Port)
 		}
+		// Ensure that pachd is really ready to receive requests.
```

Comment:
why's this check necessary? I thought that if we're able to create (& connect) a new client, we should be good to go

Reply:
Basically the proxy is being too smart.  The timeline right after deployment looks something like:

* pachd pod is ready
* k8s does a health check
* endpoints become ready
* endpoints are added to DNS
* envoy polls DNS to discover the endpoints
* envoy health checks the service
* traffic is ready to be routed

If you send a request between <pachd pod is ready> and <traffic is ready to be routed>, Envoy returns "no healthy upstream" instead of blocking until ready.  This is different than a raw TCP connection through kube-proxy.

Luckily I think disabling the health check changes the behavior to blocking until ready.  That's something I'm trying to avoid, because the message from pachctl when pachd is actually unhealthy is quite good ("no healthy upstream"), but there is some latency on transitioning from unhealthy to healthy (worst case: 2 * (connect timeout + health check interval)) and I need to get rid of that.

There a couple of options that I am testing:
1. Just don't health check.
2. Change the polling interval while unhealthy to something higher (`unhealthy_interval` https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/health_check.proto#config-core-v3-healthcheck)
3. Infer health from responses ("outlier detection": https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier#arch-overview-outlier-detection)

We will see which yields the best results.

Reply:
I talked a little about it in the description, but basically we can't avoid the "unhealthy upstream".  In our case it's not a symptom of not having a passing health check early on, but rather that DNS hasn't been updated (and polled), and thus there are 0 upstream IP addresses to route traffic to.  (The flow is: a pachd pod starts, the pod becomes ready, the k8s endpoint controller updates the endpoint, CoreDNS receives a notification from the k8s apiserver that this happened and makes the pod available in the service [because "headless service", no kube-proxy hop, rather the ready pods are returned as unique A records], and finally Envoy polls DNS and gets the address.)

I added retries inside Envoy for certain types of connection failure (when we're sure that pachd couldn't have already processed the RPCs), but unfortunately this is not a case where Envoy thinks it can retry.  (It doesn't understand that addresses arrive asynchronously and that waiting can solve the problem; someone reported a bug for this, but they worked around it and closed it.  We might want to reopen that.) 

In the end, I think this is largely a problem for our tests and not actual users.  I fixed the tests to wait for a successful RPC instead of pod readiness.  With the retries, things work very well during version upgrades / rolling restarts.  If a pachd replica is dead and Envoy hasn't noticed yet, it will have trouble opening a connection, and that is retryable.  So requests will transparently go to a new replica even when Envoy's view of the system is outdated.

(So how does this work without Envoy?  We use normal kube-proxy services and not headless services, so there is an IP address even when pachd isn't running.  It seems that the way kube-proxy sets up the network, connections will actually be delayed until there are endpoints.  So you don't really even have to wait for pod readiness to send a request, you just need to know the TCP timeout and when pachd will be ready, and any TCP connection in that interval will succeed.  Envoy, on the other hand, just says "nope" and rejects the request.  Unfortunate.

As for "why not just use a normal service and not a headless service", Envoy already introduces an extra network hop, so adding a second hop through kube-proxy seems non-ideal.  We could do that, though.)

Reply:
Ok thanks for the thorough description. I think the key detail I was missing was the "why does it work now" comparison